src/DataStructures/Tensor/Expressions/TensorExpression.hpp

```diff
@@ -405,50 +400,32 @@ struct TensorExpression<Derived, DataType, Symm, IndexList, ArgsList<Args...>> {
-  /// \return the value of the DataType of component `tensor_index`
-  template <typename... LhsIndices, typename ArrayValueType,
-            typename V = Derived,
-            Requires<tt::is_a<Tensor, V>::value> = nullptr>
-  SPECTRE_ALWAYS_INLINE type
-  get(const std::array<ArrayValueType, num_tensor_indices>& tensor_index)
-      const {
-    ASSERT(t_ != nullptr,
-           "A TensorExpression that should be holding a pointer to a Tensor "
-           "is holding a nullptr.");
-    using rhs = args_list;
-    // To deal with Tensor products we need the ordering of only the subset of
-    // tensor indices present in this term
-    using lhs = rhs_elements_in_lhs<rhs, tmpl::list<LhsIndices...>>;
-    using rhs_only_with_lhs = rhs_elements_in_lhs<lhs, rhs>;
-    using transformation = generate_transformation<rhs, lhs, rhs_only_with_lhs>;
-    return t_->get(
-        ComputeCorrectTensorIndex<transformation>::apply(tensor_index));
-  }
-
-  /// \brief return the value of type DataType with tensor index `tensor_index`
-  ///
-  /// \details
-  /// _Note:_ This version is selected if Derived is a TensorExpression
-  ///
-  /// Forward the tensor_index onwards to the next TensorExpression
-  ///
-  /// \tparam LhsIndices the tensor indices on the LHS on the expression
-  /// \tparam V used for SFINAE
   /// \param tensor_index the tensor component to retrieve
   /// \return the value of the DataType of component `tensor_index`
-  template <typename... LhsIndices, typename ArrayValueType,
-            typename V = Derived,
-            Requires<not tt::is_a<Tensor, V>::value> = nullptr>
+  template <typename... LhsIndices, typename ArrayValueType>
   SPECTRE_ALWAYS_INLINE type
   get(const std::array<ArrayValueType, num_tensor_indices>& tensor_index)
       const {
-    ASSERT(t_ == nullptr,
-           "A TensorExpression that shouldn't be holding a pointer to a "
-           "Tensor is holding one.");
-    return static_cast<const Derived&>(*this).template get<LhsIndices...>(
-        tensor_index);
+    if constexpr (tt::is_a<Tensor, Derived>::value) {
+      ASSERT(t_ != nullptr,
+             "A TensorExpression that should be holding a pointer to a Tensor "
+             "is holding a nullptr.");
+      using rhs = args_list;
+      // To deal with Tensor products we need the ordering of only the subset of
+      // tensor indices present in this term
+      using lhs = rhs_elements_in_lhs<rhs, tmpl::list<LhsIndices...>>;
+      using rhs_only_with_lhs = rhs_elements_in_lhs<lhs, rhs>;
+      using transformation =
+          generate_transformation<rhs, lhs, rhs_only_with_lhs>;
+      return t_->get(
+          ComputeCorrectTensorIndex<transformation>::apply(tensor_index));
+    } else {
+      ASSERT(t_ == nullptr,
+             "A TensorExpression that shouldn't be holding a pointer to a "
+             "Tensor is holding one.");
+      return static_cast<const Derived&>(*this).template get<LhsIndices...>(
```

Comment:
Can `static_cast<...>(*this)` be written as `(~*this)`?

Reply:
done!